### PR TITLE
Revert "Revert "Add symmetric int16 support to tflu softmax reference kernel""

### PR DESCRIPTION
**Pull request by advaitjain**
Reverts tensorflow/tensorflow#43127

#43180 removes std:vector from the reference implementation of int16 softmax so we should be all set to use it from TFLM. The continuous integration coverage has been increased with #43160 so having that pass will give us higher confidence that things are good to go.

### CODE REVIEW

**Comment by advaitjain** [comment]
@freddan80 @sicong-li-arm @yair-ehrenwald, let's try this again.

**Review by njeffrie — DISMISSED** [review]

File: tensorflow/lite/kernels/internal/common.h [hunk 1]
```diff
@@ -263,6 +263,30 @@ inline void gen_lut(const std::function<double(double)>& func, double min,
       std::min(std::max(TfLiteRound(func(max) * 32768.0), -32768.0), 32767.0);
 }
 
+// generate INT16 LUT for function(), e.g., table exp(x) and 1/(1+x) used in
+// softmax
+inline void gen_lut(const std::function<float(float)>& func, float min,
+                    float max, int16_t* table, const int num) {
```
Comment (advaitjain) [inline tensorflow/lite/kernels/internal/common.h:269]: the std::function param is resulting in dynamic memory allocation. I think changing it to be a function pointer might do the trick.

Also, please document the parameters of the function.

**Review by advaitjain — COMMENTED** [review]
The original PR needs some changes to avoid dynamic memory allocations.

I have added a note here, but feel free to open a completely new pull request into `tensorflow:master` or a pull request into this branch (`tensorflow:revert-43127-revert-38873-tflu_softnax_int16_ref`), whatever you prefer.

Use this command to reproduce the error:
```
make -f tensorflow/lite/micro/tools/make/Makefile  TARGET=bluepill kernel_softmax_test
```

File: tensorflow/lite/kernels/internal/common.h (hunk 1 above)
Comment (yair-ehrenwald) [reply tensorflow/lite/kernels/internal/common.h:269]: Yes, changing to function pointer seems to be enough.

File: tensorflow/lite/kernels/internal/common.h [hunk 2]
```diff
@@ -439,13 +471,15 @@ inline void DownScaleInt32ToInt16Multiplier(int32_t multiplier_int32_t,
...
 //  ceil(log(abs( log(2.^(0:127))+1 ))/log(2))]
 constexpr int min_log_x_output_bits(int input_bits) {
-  return input_bits > 90   ? 7
-         : input_bits > 44 ? 6
-         : input_bits > 21 ? 5
-         : input_bits > 10 ? 4
-         : input_bits > 4  ? 3
-         : input_bits > 1  ? 2
-                           : 1;
+  return input_bits > 90
```
Comment (njeffrie) [inline tensorflow/lite/kernels/internal/common.h:474]: Is this really what clang-format wants?

**Review by njeffrie — APPROVED** [review]
One quick style question - not sure this is actually something you changed.

File: tensorflow/lite/kernels/internal/common.h (hunk 2 above)
Comment (yair-ehrenwald) [reply tensorflow/lite/kernels/internal/common.h:474]: Not something I've touched. Surprisingly this is what clang-format likes 🤔
